torchtune/dev/grpo/rewards.py

```diff
@@ -5,17 +5,148 @@
 # LICENSE file in the root directory of this source tree.
 
 import re
-from typing import List, Tuple
+from typing import Dict, List, Optional, Protocol, Tuple, Union
 from xml.etree import ElementTree as ET
 
 import torch
-
-# from tensordict import TensorDict, TensorDictBase
-# from torchrl.data import Composite, TensorSpec, Unbounded
-# from torchrl.envs import Transform
+from dataclasses import dataclass, field
 
 from torchtune.modules.transforms.tokenizers import ModelTokenizer
 
+@dataclass
+class RewardOutput:
+    reward_base_name: str
+    total_reward: torch.Tensor
+    successes: torch.Tensor
+    # optional
+    rewards: Optional[Dict[str, torch.Tensor]] = field(default_factory=dict)
+
+    def log(self, prefix: str = "") -> Dict[str, float]:
+        log_dict = {}
+        prefix = f"{prefix}/{self.reward_base_name}" if prefix else self.reward_base_name
+        for reward_name, reward in self.rewards.items():
+            log_dict[f"{prefix}/{reward_name}"] = reward.mean().item()
+        log_dict[f"{prefix}"] = self.total_reward.mean().item()
+        log_dict[f"{prefix}/successes"] = self.successes.mean().item()
+        return log_dict
+
+
+class RewardBase(Protocol):
```

Comment:
I think this should just be called "Reward" without Base in the name. We haven't established a convention yet, but I don't think abstract classes should all be named base.

Reply:
I also think that this should be an abstract base class instead of a protocol. Here you're mostly using it as an abc (since you keep inheriting from it), and in general I think abc's make it easier for users who are writing custom code to trace the interface they need to implement.

Reply:
Sigh, this is such a mess on the Python side of things to be honest as there are like 95 ways of doing the same thing :-/ 

Technically, Protocols are supposed to be the most Pythonic way of going about this, but I agree that it looks cleaner if you explicitly inherit rather than duck type

Reply:
I agree that it's very unclear, and so far, torchtune has done a poor job of enforcing interfaces. Even if protocols are the future for python, I'm wary of using them since it seems their primary purpose is for type support, but what I think torchtune needs are explicit interfaces that are easy for users to implement. I think abc's are more self documenting since every example you look at in the codebase will inherit from the abc.

Reply:
Fair, I agree (FWIW we do use protocols instead of abcs for recipes :P )